oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,450 @@
+      * Documentation should not live within the code itself (e.g., in __init__.py modules).
+
+    * For system-wide documentation:
+    
+      * OEPs_ will live within the open-edx-proposals_ repo.
+      * If other system-wide documentation is needed, they will live within the edx-developer-docs_ repo.
+
+  * **Exported Diagrams**. In order to keep diagrams updated across time, they will need to be exported
+    from whichever tool they were created in, with the exported version maintained in Github.
+
+    * draw.io diagrams will be exported and imported as XML files.
+    * Lucidchart diagrams will be exported and imported as Visio (VDX) files.
+
+* **Version Controlled**. Documentation must be maintained under version control in Github, for the
+  following reasons:
+
+  * **Reviewed**. They will be properly reviewed as defined by that repo's process. Anyone in the Open edX
+    community with access to the repo will have the ability to add, update, fix, and contribute to
+    documentation by following the repo's review and maintainence process.
+
+  * **Tagged**. They will be appropriately in-synch with their corresponding code. For example,
+    when an open edX instance runs an older named release, the operator can refer to the matching
+    documentation for that version of the codebase.
+
+  * **Source Controlled**. They will have the advantages of having source-control backed files, including sharing
+    across multiple developers and robust storage.
+
+.. _edx-developer-docs: https://github.com/edx/edx-developer-docs
+.. _docs.edx.org: https://docs.edx.org/
+
+Format
+======
+
+We will use `reStructuredText (rST)`_ for all developer documentation.
+
+* rST scales up: it can be used for anything from docstrings to single-page
+  documents to books to multi-volume book sets.
+
+* rST has a rich feature set, including inter-document links, table of
+  contents, auto-collection of code documentation, list-formatted tables,
+  semantic markup, and multiple output formats.
+
+* A single documentation format enables consistency.  It also avoids conversion
+  to rST when we require its additional features.  Our experience has shown
+  that these two benefits of standardization keep down the overall cost of
+  documentation maintenance.
+
+* rST is the standard markup format for Python project documentation.
+
+* rST does not suffer from incompatible implementations, with different
+  features supported differently by different renderers.
+
+* GitHub renders rST natively.
+
+More details on these points are in Eric Holscher's `why rST over Markdown for
+documentation`_ blog post.
+
+.. _why rST over Markdown for documentation: http://ericholscher.com/blog/2016/mar/15/dont-use-markdown-for-technical-docs/
+
+
+Best practices
+--------------
+
+* We strongly recommend keeping line length between 80 and 120 characters. Keeping it less than 120
+  leaves room for future edits.
```

Comment:
Why?  Paragraphs of text are much more readable, let alone maintainable, when the viewer wraps them to the view width; I would argue that best-practice in documentation is strongly **do not** manually wrap paragraphs.

Reply:
Except if the paragraph is in a list of some sort, so the auto-wrapping will get the indent wrong.

Keeping an entire paragraph on one line makes diffs difficult.

Reply:
Hard-wrapped paragraphs are ***much*** worse when they have to auto-wrap anyway; if a paragraph is in a list the indent may be wrong (if the viewer is not format-aware) but at least it's consistent, but if the text was hard-wrapped, it's wrong _and_ fragmented and pretty much unreadable.

Good diff tools highlight changed portions within the diffs, we can come up with some recommendations.

Reply:
I guess we haven't discussed assumptions about how people will read these docs.  I believe people will either:
- In their plain-text editor of choice, and we can assume that they are at least 80 characters wide,
- Or viewing a rendered document, for example, browsing GitHub.

For a rendered document, the wrapping choice doesn't matter, the result is the same.

Reply:
The assumption appears to be 120-character minimum width, not 80.  Nevertheless, an unstated assumption hardly seems like a strong basis for a strong recommendation, especially for something that arguably harms both readability and maintainability.

Reply:
A default editor window with my settings with a files view along the side is *barely* 80 characters wide; full-screen without the files view is 128-characters; if I open another document alongside or the reStructured Text preview pane, the editor is about 60 characters, still full-screen, no files view.

Reply:
Without hard-wrapping, text is consistent across all of these; with manual line breaks it looks terrible at everything except full-screen-full-width and will be a pain to edit as well.

Reply:
@inventhouse:
1. I think you can just drag an image here to attach it if you want.
2. I'm curious how you deal with docstrings and code line length limits that these recommendations mimic.

Reply:
@robrap: To your second question, I'm strongly opposed to line-length limits in code as well, more so in fact:

- They're arbitrary and sub-optimal for any width other than the "chosen" size
- They're annoying to maintain (*Especially* at narrower/smaller screen sizes)
- They're bad for readability, auto-wraping is much more consistent over a range of screen and font sizes; modern editors even do a pretty good job of matching indentation level for wrapped lines
- In code they create perverse incentives against self-documenting names and use of named arguments
- There are *good* reasons to keep lines of *code* short and break logic into steps, but simplistic character count is a poor proxy for complexity.
- They add considerable friction to writing substantial comments or documentation for negative benefit

Reply:
I manually un-manually-wrapped one of the lists in this document, the one on the right is *more* readable and *easier* to edit and maintain than the one on the left:

![image](https://user-images.githubusercontent.com/17170027/51003132-9b885880-1503-11e9-8c65-5e3105e709a9.png)

I repeat my first question: **Why?**

Reply:
Also, given the limitations with paragraph diffing on GitHub are there any recommended doc-diffing tools we can use/integrate with in the meantime, while GitHub improves its own tool?

Reply:
Great!  As for diff tools, they are myriad and I'd love to get all the recommendations we can; one that I've used a bit is the git-time-machine plugin for Atom: https://atom.io/packages/git-time-machine

![image](https://user-images.githubusercontent.com/17170027/51037780-2eb2a400-157f-11e9-9c94-775dcf6cd7bd.png)

Reply:
I updated the OEP to recommend auto-wrap over fixed line lengths.  Please review.

Reply:
I like that the new recommendation includes a summary of the reasoning 👍